File: Convolution/Blur.cpp
#include <opencv2/opencv.hpp>
#include <iostream>

using namespace std;
using namespace cv;

int main(int argc, const char *argv[])
{
	Mat src = imread("D:/Images/lena.jpg");
	if (src.empty())
	{
		printf("could not load image...\n");
		return -1;
	}
	namedWindow("src", WINDOW_AUTOSIZE);
	imshow("src", src);

	//blur 
	Mat dst = src.clone();
	//为了使卷积核能够重合,i和j都从1开始遍历，到rows-1和cols-1
	for (int i = 1; i < src.rows - 1; i++)
	{
		for (int j = 1; j < src.cols - 1; j++)
		{
			Vec3b p1 = src.at<Vec3b>(i - 1, j - 1);
			Vec3b p2 = src.at<Vec3b>(i - 1, j);
			Vec3b p3 = src.at<Vec3b>(i - 1, j + 1);
			Vec3b p4 = src.at<Vec3b>(i, j - 1);
			Vec3b p5 = src.at<Vec3b>(i, j);
			Vec3b p6 = src.at<Vec3b>(i, j + 1);
			Vec3b p7 = src.at<Vec3b>(i + 1, j - 1);
			Vec3b p8 = src.at<Vec3b>(i + 1, j);
			Vec3b p9 = src.at<Vec3b>(i + 1, j + 1);

			int b = p1[0] + p2[0] + p3[0] + p4[0] + p5[0] + p6[0] + p7[0] + p8[0] + p9[0];
			int g = p1[1] + p2[1] + p3[1] + p4[1] + p5[1] + p6[1] + p7[1] + p8[1] + p9[1];
			int r = p1[2] + p2[2] + p3[2] + p4[2] + p5[2] + p6[2] + p7[2] + p8[2] + p9[2];
			dst.at<Vec3b>(i, j)[0] = saturate_cast<uchar>(b / 9);
			dst.at<Vec3b>(i, j)[1] = saturate_cast<uchar>(g / 9);
			dst.at<Vec3b>(i, j)[2] = saturate_cast<uchar>(r / 9);
		}
	}
	imshow("blur", dst);
	imwrite("D:/Images/bulrRes.png", dst);

	//Opencv实现
	Mat result;
	//Size
	//使用积分图技术加速
	blur(src, result, Size(15, 15), Point(-1, -1), 4);
	imshow("result", result);

	waitKey(0);
	return 0;

}

File: Convolution/Blur.py
import cv2 as cv
import numpy as np

src = cv.imread("D:/Images/lena.jpg")
cv.namedWindow("src", cv.WINDOW_AUTOSIZE)
cv.imshow("src", src)


def custom_blur_avg(src):
    width, height, channel = src.shape
    dst = np.copy(src)
    for i in range(1, height - 1, 1):
        for j in range(1, width - 1, 1):
            t1 = np.int32(src[i - 1, j - 1])
            t2 = np.int32(src[i - 1, j])
            t3 = np.int32(src[i - 1, j + 1])
            t4 = np.int32(src[i, j - 1])
            t5 = np.int32(src[i, j])
            t6 = np.int32(src[i, j + 1])
            t7 = np.int32(src[i + 1, j - 1])
            t8 = np.int32(src[i + 1, j])
            t9 = np.int32(src[i + 1, j + 1])

            b = t1[0] + t2[0] + t3[0] + t4[0] + t5[0] + t6[0] + t7[0] + t8[0] + t9[0]
            g = t1[1] + t2[1] + t3[1] + t4[1] + t5[1] + t6[1] + t7[1] + t8[1] + t9[1]
            r = t1[2] + t2[2] + t3[2] + t4[2] + t5[2] + t6[2] + t7[2] + t8[2] + t9[2]
            dst[i, j] = [b // 9, g // 9, r // 9]
    return dst


dst = custom_blur_avg(src)
cv.imshow("dst", dst)
dst1 = np.copy(src)
dst1 = cv.blur(src, ksize=(15, 15), anchor=(-1, -1), borderType=4)
cv.imshow("dst1", dst1)

cv.waitKey(0)
cv.destroyAllWindows()
